extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/data/input/kafka/KafkaTopicPartition.java

```diff
@@ -0,0 +1,183 @@
+import java.io.IOException;
+import java.util.Objects;
+import java.util.Optional;
+
+/**
+ * This class represents the partition id for kafka ingestion. This partition id includes topic name along with an
+ * integer partition. The topic name is required because the same partition id can be used for different topics.
+ * This class is used as a key in {@link org.apache.druid.indexing.kafka.KafkaDataSourceMetadata} to store the offsets
+ * for each partition.
+ *
+ */
+@JsonSerialize(using = KafkaTopicPartition.KafkaTopicPartitionSerializer.class,
+    keyUsing = KafkaTopicPartition.KafkaTopicPartitionSerializer.class)
+@JsonDeserialize(using = KafkaTopicPartition.KafkaTopicPartitionDeserializer.class, keyUsing =
+    KafkaTopicPartition.KafkaTopicPartitionKeyDeserializer.class)
+public class KafkaTopicPartition
+{
+  private final int partition;
+  @Nullable
+  private final String topic;
+
+  // This flag is used to maintain backward incompatibilty with older versions of kafka indexing. If this flag
+  // is set to false,
+  // - KafkaTopicPartition will be serialized as an integer and can be read back by older version.
+  // - topic field is ignored while comparing two KafkaTopicPartition objects and calculating hashcode.
+  // This flag must be explicitly passed while constructing KafkaTopicPartition object. That way, we can ensure that
+  // a particular supervisor is always running in multi topic mode or single topic mode.
+  private final boolean multiTopicPartition;
+
+  public KafkaTopicPartition(boolean multiTopicPartition, @Nullable String topic, int partition)
+  {
+    this.partition = partition;
+    this.topic = topic;
+    this.multiTopicPartition = multiTopicPartition;
+  }
+
+  public int partition()
+  {
+    return partition;
+  }
+
+  public Optional<String> topic()
+  {
+    return Optional.ofNullable(topic);
+  }
+
+  public boolean isMultiTopicPartition()
+  {
+    return multiTopicPartition;
+  }
+
+  public TopicPartition asTopicPartition(String fallbackTopic)
+  {
+    return new TopicPartition(topic != null ? topic : fallbackTopic, partition);
+  }
+
+  @Override
+  public String toString()
+  {
+    // TODO - fix this so toString is not used for serialization
+    if (null != topic && multiTopicPartition) {
+      return topic + ":" + partition;
+    } else {
+      return Integer.toString(partition);
+    }
+  }
+
+  @Override
+  public boolean equals(Object o)
+  {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    KafkaTopicPartition that = (KafkaTopicPartition) o;
+    return partition == that.partition && multiTopicPartition == that.multiTopicPartition && (!multiTopicPartition
+                                                                                              || Objects.equals(
+        topic,
+        that.topic
+    ));
+  }
+
+  @Override
+  public int hashCode()
+  {
+    if (multiTopicPartition) {
```

Comment:
Hmm. Good point. Its just a defensive check that if someone is passing a topic for a single-topic mode, we don't compare. I was seeing some test failures because of that. But I realize now that it's entirely in my control, how those objects get created. I will simplify these equals/hashcode methods. But in the constructor, I will set the topic to null if multi-topic mode is not enabled. what do you think?